rfd/XXXX-cloud-agent-upgrades.md

```diff
@@ -0,0 +1,318 @@
+# (i.e. this is the "repeating" portion of the timer)
+OnUnitActiveSec=12h
+
+# randomized delay sec adds an added delay to all firings
+RandomizedDelaySec=90
+
+[Install]
+WantedBy=teleport-agent.service
+```
+
+The exact time values are TBD. We want upgrades to be reliably and frequently discovered, but don't want to end up overloading
+repositories. This is why we aren't just using `OnActiveSec=1`.  If teleport is in *too* tight of a crash loop, we don't want
+to just end up hitting the repos in a tight loop. This will definitely take some fine-tuning, and we may want to customize the
+`StartLimitIntervalSec` and `StartLimitBurst` of the main teleport agent unit.
+
+---
+
+TODO: apt/yum channel and artifact promotion details
+
+
+### Automatic/Guided Install Changes
+
+Teleport provides two kinds of automatic/guided install mechanism for adding the teleport agent software to existing
+machines. There are "discovery installers" which are scripts used by ec2 auto-discovery (and possibly by other forms of
+server auto-discovery in the future) to install teleport agents automatically, and there is the `node-join` script template
+used by the web UI to serve custom installers during the interactive agent-adding process.
+
+To ensure a seamless experience, we will need both of these install scripts to be capable of being parameterized or overridden s.t.
+they target `cloud-stable` when the cluster is configured for continuous agent upgrades.
+
+The current implementation of the `node-join` script template relies upon direct downloads from `get.gravitational.com`. It would
+be preferable to force use of a package manager when installing `cloud-stable`, as that would let us improve standardization, minimize
+the amount of custom logic in the download path, and (most importantly) reduce the complexity of the periodic upgrade operation,
+which is the most sensitive part of this proposal.
+
+I believe the best possible outcome would be to unify the core logic of the "discovery", "join", and "upgrade" scripts into a single
+cross-platform function so that we have a single source of truth for package manager discovery and invocation going forward. The
+feasability of this will require further investigation.
+
+
+### Summary of Changes
+
+- Agents will expose a flags to enable the VM and k8s variants of the agent upgrade behavior.
+
+- Package repositories will expose a `cloud-stable` variant of packages which will lag slightly behind "latest"
+and be managed indipendently of the standard release process.
+
+- A new `cloud-stable` helm chart (or option) will be dstributed.
+
+- A new `teleport-upgrade` script and accompanying systemd timer unit will be distributed as part of the
+`cloud-stable` apt/yum packages.
+
+- Teleport auth servers will be modified to be able to serve the maintenance window to agents.
+
+- Cloud clusters will have the option to track `cloud-stable` rather than a specific major version.
```

Comment:
This comment applies to several place in the RFD however I thought it'd be best to group everything together in one place for discussion. I would like to propose an alternative to using a `cloud-stable` OS package repo (apt/yum) channel as a source of truth for cloud update version. Instead, I propose that we setup a S3 bucket that contains cloud update information. This has the following advantages:
* We don't need to re-publish our OS packages to another repo. We would continue to use the repos/repo components that we use for all other forms of installation.
* Rollbacks are easy. This is a separate discussion, but rollbacks on an OS package repository are difficult for a number of reasons, and have a lot of places for potential screw-ups. These screw-ups can happen in the context of the OS package repository, or (more likely) on each individual node.
* In the future, should the updater script need more information than the version it should install (i.e. more than "v1.2.3"), it would be very easy to add that information compared to apt/yum repos.
* When something breaks during package update deployment, and requires manual intervention, it is _significantly_ easier to make changes to a simple text file than an entire apt/yum repository.
* Implementing and maintaining this approach is a little simpler from an internal tooling perspective. There are less moving parts (`s3 copy` vs a medium-size go tool I wrote for OS package repo management) so things are less likely to break.
* This bucket could potentially be used for other automatic upgrade information in the future as it is independent of specific packaging systems.

To implement this we would setup a new S3 bucket fronted by CloudFront at `https://update.gravitational.io` or similar. The updater script would `VERSION=$(curl https://update.gravitational.io/cloud-version)` (or maybe `curl https://update.gravitational.io/v1/cloud-version` as versioning is always good), and then `apt install -y teleport==$VERSION`.

Alternatively this functionality could be added to the release service, which would likely take longer to implement but may be a more appropriate place.

Reply:
Updated the proposal to include the basic outline of this plan.